firmware/src/opaygo_functions.hpp

```diff
@@ -0,0 +1,313 @@
+    if (TokenValue == COUNTER_SYNC_VALUE) {
+      BlinkGreenLED(
+          3,
+          BLINK_PERIOD);  // We blink green twice to show that the token is good
+    } else if (TokenValue == PAYG_DISABLE_VALUE) {
+      PAYGEnabled = false;
+      BlinkGreenLED(5, BLINK_PERIOD);  // We blink green twice to show that the
+                                       // device is active forever
+    } else {
+      if (ActivationCount % 2) {
+        PAYGEnabled = true;
+        SetTime(TokenValue);
+      } else {
+        AddTime(TokenValue);
+      }
+      BlinkGreenLED(
+          2,
+          BLINK_PERIOD);  // We blink green twice to show that the token is good
+    }
+    StoreActivationVariables();  // We store in every case
+  }
+}
+
+bool IsActive() {
+  if (PAYGEnabled) {
+    if (ActiveUntil > GetTimeInSeconds()) {
+      return true;
+    } else {
+      return false;
+    }
+  } else {
+    return true;
+  }
+}
+
+uint64_t WaitForTokenEntry() {
+  uint64_t TempToken = 0;
+  bool NoToken = true;
+  int LastKey;
+
+  while (NoToken) {
+    LastKey = GetKeyPressed();
+    if (LastKey == STAR_KEY) {
+      if (TokenEntryAllowed()) {
+        NoToken = false;
+      } else {
+        BlinkRedLED(1, BLINK_PERIOD);
+#ifdef DEBUG
+        printf("\nToken entry locked for %" PRIu32 "seconds",
+               TokenEntryLockedUntil - GetTimeInSeconds());
+#endif
+      }
+    } else if (LastKey == HASH_KEY) {
+      if (IsActive()) {
+        BlinkGreenLED(1, BLINK_PERIOD);
+#ifdef DEBUG
+        printf("\nTime Left: %" PRIu32 "seconds",
+               ActiveUntil - GetTimeInSeconds());
+#endif
+      } else {
+        BlinkRedLED(1, BLINK_PERIOD);
+      }
+    }
+  }
+  for (int i = 0; i < TOKEN_LENGTH; i++) {
+    // We add the last key pressed to the token (as integer) if needed
+    TempToken += GetKeyPressed() * pow(10, (TOKEN_LENGTH - 1) - i);
+  }
+  return TempToken;
+}
+
+void initializeTime() {
+  if (!rtc.begin()) {
+    Serial.println("Couldn't find RTC");
+    while (1)
+      ;
+  }
+
+  if (!rtc.isrunning()) {
+    Serial.println("RTC is NOT running!");
+  }
+
+  // We check that it is not a disconnection power - Arduino
+  uint32_t nvramCheck = readUint32FromNvram(TIME_INITIALIZATION_NVRAM_ADDRESS);
+  if (nvramCheck == 0) {  // it is the first setup of the Arduino
+    rtc.adjust(
+        DateTime(F(__DATE__), F(__TIME__)));  // sets the RTC to the date & time
+                                              // this sketch was compiled
+    DateTime now = rtc.now();
+    timeInitializationRtc = now.unixtime();
+    storeUint32InNvram(timeInitializationRtc,
+                       TIME_INITIALIZATION_NVRAM_ADDRESS);
+  } else {  // a disconnection happened
+    timeInitializationRtc = nvramCheck;
+    incrementNbDisconnectionsEeprom();
+    storeTimeStampEEPROM(
+        GetTimeInSeconds());  // to avoid counting twice the disconnection if
+                              // the previous TimeStamp was too long ago
+  }
+}
```

Comment:
This isn't PAYGO related, but time related, should it go in `time_management.hpp` instead?